homeassistant/components/mqtt/siren.py

```diff
@@ -0,0 +1,341 @@
+            """Handle new MQTT state messages."""
+            payload = self._value_template(msg.payload)
+            if payload == self._state_on:
+                self._attr_is_on = True
+            elif payload == self._state_off:
+                self._attr_is_on = False
+
+            self.async_write_ha_state()
+
+        if self._config.get(CONF_STATE_TOPIC) is None:
+            # Force into optimistic mode.
+            self._optimistic = True
+        else:
+            self._sub_state = await subscription.async_subscribe_topics(
+                self.hass,
+                self._sub_state,
+                {
+                    CONF_STATE_TOPIC: {
+                        "topic": self._config.get(CONF_STATE_TOPIC),
+                        "msg_callback": state_message_received,
+                        "qos": self._config[CONF_QOS],
+                        "encoding": self._config[CONF_ENCODING] or None,
+                    }
+                },
+            )
+
+    @property
+    def assumed_state(self):
+        """Return true if we do optimistic updates."""
+        return self._optimistic
+
+    @property
+    def supported_features(self) -> int:
+        """Flag supported features."""
+        return self._supported_features
+
+    async def async_conditional_publish(
+        self, topic: str, template: str, value: Any, variables: dict | None = None
+    ) -> None:
+        """Publish MQTT payload with command template if a topic is configured."""
+        if self._config.get(topic) and value:
+            payload = self._command_templates[template](value, variables)
+
+            await mqtt.async_publish(
+                self.hass,
+                self._config[topic],
+                payload,
+                self._config[CONF_QOS],
+                self._config[CONF_RETAIN],
+                self._config[CONF_ENCODING],
+            )
+
+    async def async_turn_on(self, **kwargs) -> None:
+        """Turn the siren on.
+
+        This method is a coroutine.
+        """
+        await self.async_conditional_publish(
+            CONF_COMMAND_TOPIC,
+            CONF_COMMAND_TEMPLATE,
+            self._config[CONF_PAYLOAD_ON],
+            kwargs,
+        )
+        if self._optimistic:
+            # Optimistically assume that siren has changed state.
+            self._attr_is_on = True
+            self.async_write_ha_state()
+
+        if not kwargs:
+            return
+
+        await self.async_conditional_publish(
+            CONF_DURATION_COMMAND_TOPIC,
+            CONF_DURATION_COMMAND_TEMPLATE,
+            kwargs.get(ATTR_DURATION),
+        )
+        await self.async_conditional_publish(
+            CONF_TONE_COMMAND_TOPIC,
+            CONF_TONE_COMMAND_TEMPLATE,
+            kwargs.get(ATTR_TONE),
+        )
+        await self.async_conditional_publish(
+            CONF_VOLUME_COMMAND_TOPIC,
+            CONF_VOLUME_COMMAND_TEMPLATE,
+            kwargs.get(ATTR_VOLUME_LEVEL),
+        )
```

Comment:
This doesn't seem ideal, wouldn't it be better to send the options in the same message, otherwise we turn on the siren first with wrong volume and tone?
Is there some external package which supports controlling sirens over MQTT, does zigbee2mqtt support it for example?

Reply:
I changed the whole set up. Left one command and state topic with json support.